### PR TITLE
simplify app identifier management

### PR DESCRIPTION
**Pull request by dsokal**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary.

# Why

This is part of the preparation for supporting managed credentials for multitarget iOS projects.
Application id management is currently quite complicated. For generic projects, EAS CLI allows for synchronization between app.json and native code. There are a lot of corner cases like what to do if there are a few targets with different bundle ids. If we want to support iOS projects with multiple targets (and so multiple bundle identifiers) we'd have to add even more complexity to the code.

After a few discussions with @wkozyra95, we decided that this is a good time to simplify the app identifier management so that:
- For managed projects, we rely on the value of `ios.bundleIdentifier`/`android.package` from app.json/app.config.js.
- For generic projects, we rely only on the value defined in the Xcode/Android native project. If either the bundle id or Android package is defined in the app config, it's ignored and we print a warning that it's deprecated. 

Generally, this lets us keep parity with the current implementation with the exception of one thing. It won't be possible to change the bundle id/application id in native code just by modifying the app config and then running an EAS Build. However, I think if someone decides to go for a bare project, they should be able to change those values on their own (with Xcode/Android Studio). We could improve the experience by adding a prompt to `expo init` to configure bundle id/application id when initing a bare project. For managed, it's not necessary as we'll prompt for those values when running a build.

This PR also addresses https://linear.app/expo/issue/ENG-894/automatically-add-bundleidentifier-and-package-when-running-eas-build. When a managed project has been just inited with Expo CLI, it should be possible to build it without having to manually edit `app.json` file to add bundle identifier.

# How

- I removed the logic responsible for synchronizing bundle identifier and application id (android package) between app.json and native code. 
- I removed the `experimental.disableIosBundleIdentifierValidation` flag from eas.json as it's not relevant anymore. 
- I added printing a warning when `ios.bundleIdentifier`/`android.package` is defined in app.json but the project is generic. 

# Test Plan

- TODO: add unit tests for functions in `bundleIdentifier.ts`/`applicationId.ts`
- Tested manually by building Android and iOS builds for a managed and generic projects.

### CODE REVIEW

File: packages/eas-cli/src/build/android/applicationId.ts [hunk 1]
```diff
@@ -1,191 +1,114 @@
...
+  }
+
+  const { packageName } = await promptAsync({
+    name: 'packageName',
+    type: 'text',
+    message: `What would you like your Android package name to be?`,
+    validate: value => (isApplicationIdValid(value) ? true : INVALID_APPLICATION_ID_MESSAGE),
+  });
 
   const rawStaticConfig = await fs.readJSON(paths.staticConfigPath);
```
Comment (wkozyra95) [inline packages/eas-cli/src/build/android/applicationId.ts:86]: use https://github.com/expo/expo-cli/blob/master/packages/config/src/Config.ts#L370

File: packages/eas-cli/src/build/android/applicationId.ts [hunk 2]
```diff
@@ -1,191 +1,114 @@
...
+  if (!paths.staticConfigPath) {
+    throw new Error(
+      `"android.package" is not defined in your app.config.js and we can't update this file programatically. Add the value on your own and run this command again.`
+    );
+  }
+
+  const { packageName } = await promptAsync({
+    name: 'packageName',
+    type: 'text',
+    message: `What would you like your Android package name to be?`,
```
Comment (wkozyra95) [inline packages/eas-cli/src/build/android/applicationId.ts:82]: android package is not really correct, we use field `package` in app.json but it's applicationId we want to define here

File: packages/eas-cli/src/build/build.ts [hunk 3]
```diff
@@ -103,6 +103,12 @@ export async function prepareBuildRequestForPlatformAsync<
     projectConfiguration: builder.projectConfiguration,
   });
 
+  let x = 1;
+  x = 2;
+  if (x === 2) {
+    process.exit(2137);
+  }
+
```
Comment (wkozyra95) [inline packages/eas-cli/src/build/build.ts:111]: ?

File: packages/eas-cli/src/build/ios/bundleIdentifier.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  const { bundleIdentifier } = await promptAsync({
+    name: 'bundleIdentifier',
+    type: 'text',
+    message: `What would you like your iOS bundle identifier to be?`,
+    validate: value => (isBundleIdentifierValid(value) ? true : INVALID_BUNDLE_IDENTIFIER_MESSAGE),
+  });
+
+  const rawStaticConfig = await fs.readJSON(paths.staticConfigPath);
```
Comment (wkozyra95) [inline packages/eas-cli/src/build/ios/bundleIdentifier.ts:81]: use https://github.com/expo/expo-cli/blob/master/packages/config/src/Config.ts#L370

File: packages/eas-cli/src/credentials/ios/actions/BuildCredentialsUtils.ts [hunk 5]
```diff
@@ -1,13 +1,16 @@
+import { Platform } from '@expo/eas-build-job';
 import nullthrows from 'nullthrows';
 
+import { getBundleIdentifier } from '../../../build/ios/bundleIdentifier';
+import { resolveWorkflow } from '../../../build/utils/workflow';
```
Comment (wkozyra95) [inline packages/eas-cli/src/credentials/ios/actions/BuildCredentialsUtils.ts:5]: code outside `build` directory should not be aware of where bundleIdentifier is located, this resolve function should be part of the getBundleIdentifier

Also, I think the function for accessing bundleIdentifier should not be in the `build` directory

**Review by wkozyra95 — APPROVED** [review]

**Comment by brentvatne** [comment]
> For generic projects, we rely only on the value defined in the Xcode/Android native project. If either the bundle id or Android package is defined in the app config, it's ignored and we print a warning that it's deprecated.

:+1:

> For managed projects, we rely on the value of ios.bundleIdentifier/android.package from app.json/app.config.js.

one thing we should consider here (it may not actually impact this particular pr) is that in the future we will want to support multiple bundle identifiers for managed projects too. it's not exactly clear how this will work at this time, but we'll want to try to figure it out at some point.

> However, I think if someone decides to go for a bare project, they should be able to change those values on their own (with Xcode/Android Studio).

agreed

> We could improve the experience by adding a prompt to expo init to configure bundle id/application id when initing a bare project.

i discussed this briefly with @EvanBacon and i think we won't add this right away, but we're thinking of moving towards a single default template in `expo init` and possibly prompting if they want it to access the native projects, then reusing `prebuild` which already prompts for bundleIdentifier/package. so, no need to follow up from this pr with an implementation for this.

**Comment by actuallymentor** [comment]
Ok, so this change caught me completely off guard.

> However, I think if someone decides to go for a bare project, they should be able to change those values on their own (with Xcode/Android Studio)

I can confirm this is not the case.

I've used and loved Expo for years, and have never touched Xcode or Android studio apart from opening a simulator. The only reason I am ejected is because react native Firebase requires native additions.

May I recommend **at least** adding a `learn more` link in the CLI deprecation method?

**Comment by dsokal** [comment]
@actuallymentor Could you please open an issue?

**Pull request merged**